src/stream/rtsp_server.rs

```diff
@@ -0,0 +1,169 @@
+
+            // Attach the server to our main context.
+            // A main context is the thing where other stuff is registering itself for its
+            // events (e.g. sockets, GStreamer bus, ...) and the main loop is something that
+            // polls the main context for its events and dispatches them to whoever is
+            // interested in them. In this example, we only do have one, so we can
+            // leave the context parameter empty, it will automatically select
+            // the default one.
+            let id = rtsp_server.server.attach(None).unwrap();
+
+            // Start the mainloop. From this point on, the server will start to serve
+            // our quality content to connecting clients.
+            let main_loop = glib::MainLoop::new(None, false);
+            rtsp_server.is_running = true;
+            drop(rtsp_server);
+
+            main_loop.run();
+
+            let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
+            rtsp_server.is_running = false;
+            drop(rtsp_server);
+
+            id.remove();
+        }
+    }
+
+    pub fn add_pipeline(pipeline_description: &str, path: &str) {
+        // Initialize the singleton before calling gstreamer factory
+        let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
+
+        let factory = gstreamer_rtsp_server::RTSPMediaFactory::new();
+        factory.set_launch(&pipeline_description);
+        factory.set_shared(true);
+
+        // TODO: handle debug/errors:
+        rtsp_server
+            .path_to_factory
+            .insert(path.to_string(), factory);
+    }
+
+    #[allow(dead_code)]
+    pub fn remove_pipeline(path: &str) {
+        let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
+        // TODO: handle debug/errors:
+        rtsp_server.path_to_factory.remove(path);
+    }
+
+    pub fn server_start(rtsp_server_host: &str, rtsp_server_port: u16) {
+        let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
+        if rtsp_server.is_running {
+            return;
+        }
+        rtsp_server.host = rtsp_server_host.to_string();
+        rtsp_server.port = rtsp_server_port;
+
+        rtsp_server.server.set_address(&rtsp_server.host);
+
+        rtsp_server
+            .server
+            .set_service(&rtsp_server.port.to_string());
+    }
+
+    pub fn start(path: &str) {
+        let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
+
+        // Much like HTTP servers, RTSP servers have multiple endpoints that
+        // provide different streams. Here, we ask our server to give
+        // us a reference to his list of endpoints, so we can add our
+        // test endpoint, providing the pipeline from the cli.
+        let mounts = rtsp_server
+            .server
+            .mount_points()
+            .ok_or("Could not get mount points")
+            .unwrap();
+
+        let factory = &rtsp_server.path_to_factory[path];
```

Comment:
This access does the unwrap for you, we should replace it or we should only run with a previous check to make sure that path exist inside the hashmap.